File: tools/communication/openrouter/src/python/openrouter_tool.py
import json
from typing import Dict, Any, Optional, List
from bridges.python.src.sdk.base_tool import BaseTool
from bridges.python.src.sdk.toolkit_config import ToolkitConfig
from bridges.python.src.sdk.network import Network, NetworkError

# Hardcoded default settings for OpenRouter tool
OPENROUTER_API_KEY = None
OPENROUTER_MODEL = "google/gemini-3-flash-preview"
DEFAULT_SETTINGS = {
    "OPENROUTER_API_KEY": OPENROUTER_API_KEY,
    "OPENROUTER_MODEL": OPENROUTER_MODEL,
}
REQUIRED_SETTINGS = ["OPENROUTER_API_KEY"]


class OpenRouterTool(BaseTool):
    """OpenRouter tool for unified LLM API access across all skills"""

    TOOLKIT = "communication"

    def __init__(self, api_key: Optional[str] = None):
        super().__init__()
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)

        tool_settings = ToolkitConfig.load_tool_settings(
            self.TOOLKIT, self.tool_name, DEFAULT_SETTINGS
        )
        self.settings = tool_settings
        self.required_settings = REQUIRED_SETTINGS
        self._check_required_settings(self.tool_name)

        # Priority: skill-provided api_key > toolkit settings > hardcoded default
        self.api_key = api_key or self.settings.get(
            "OPENROUTER_API_KEY", OPENROUTER_API_KEY
        )

        # Load model settings
        self.model = self.settings.get("OPENROUTER_MODEL", OPENROUTER_MODEL)

        self.network = Network({"base_url": "https://openrouter.ai/api"})

    @property
    def tool_name(self) -> str:
        return "openrouter"

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config["description"]

    def set_api_key(self, api_key: str) -> None:
        """Set the OpenRouter API key"""
        self.api_key = api_key

    def chat_completion(
        self,
        messages: List[Dict[str, str]],
        model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        system_prompt: Optional[str] = None,
        use_structured_output: bool = False,
        json_schema: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Send a chat completion request to OpenRouter

        Args:
            messages: List of message dictionaries with 'role' and 'content'
            model: Model ID (full OpenRouter model ID, e.g. 'google/gemini-3-flash-preview')
            temperature: Sampling temperature (0-2)
            max_tokens: Maximum tokens to generate
            system_prompt: System prompt to prepend
            use_structured_output: Whether to use OpenRouter's structured outputs
            json_schema: JSON schema for structured output (required if use_structured_output=True)

        Returns:
            Dict with response data or error information
        """
        if not self.api_key:
            return {"success": False, "error": "OpenRouter API key not configured"}

        # Use default model if none provided
        model = model or self.model

        # Prepare messages with system prompt if provided
        request_messages = []
        if system_prompt:
            request_messages.append({"role": "system", "content": system_prompt})
        request_messages.extend(messages)

        # Prepare request payload
        payload = {
            "model": model,
            "messages": request_messages,
            "temperature": temperature,
        }

        if max_tokens:
            payload["max_tokens"] = max_tokens

        # Add structured output configuration if requested
        if use_structured_output and json_schema:
            payload["response_format"] = {
                "type": "json_schema",
                "json_schema": {
                    "name": json_schema.get("name", "response"),
                    "strict": True,
                    "schema": json_schema["schema"],
                },
            }

        try:
            response = self.network.request(
                {
                    "url": "/v1/chat/completions",
                    "method": "POST",
                    "headers": {
                        "Authorization": f"Bearer {self.api_key}",
                        "Content-Type": "application/json",
                    },
                    "data": payload,
                }
            )

            return {"success": True, "data": response["data"], "model_used": model}

        except NetworkError as e:
            return {
                "success": False,
                "error": f"OpenRouter API error: {str(e)}",
                "status_code": getattr(e.response, "status_code", None),
            }

    def completion(
        self,
        prompt: str,
        model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        system_prompt: Optional[str] = None,
        use_structured_output: bool = False,
        json_schema: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        General text completion for any use case

        Args:
            prompt: Text prompt to complete
            model: Model ID (full OpenRouter model ID)
            temperature: Sampling temperature
            max_tokens: Maximum tokens to generate
            system_prompt: Optional system prompt
            use_structured_output: Whether to use structured outputs
            json_schema: JSON schema for structured output

        Returns:
            Dict with completion result
        """
        messages = [{"role": "user", "content": prompt}]

        response = self.chat_completion(
            messages=messages,
            model=model,
            temperature=temperature,
            max_tokens=max_tokens,
            system_prompt=system_prompt,
            use_structured_output=use_structured_output,
            json_schema=json_schema,
        )

        if not response["success"]:
            return response

        try:
            content = response["data"]["choices"][0]["message"]["content"]
            return {
                "success": True,
                "content": content,
                "model_used": response["model_used"],
            }
        except (KeyError, IndexError) as e:
            return {
                "success": False,
                "error": f"Failed to extract completion: {str(e)}",
            }

    def structured_completion(
        self,
        prompt: str,
        json_schema: Dict[str, Any],
        model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        system_prompt: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Generate structured JSON output using OpenRouter's structured outputs feature

        Args:
            prompt: Text prompt to complete
            json_schema: JSON schema defining the required output structure
            model: Model ID (full OpenRouter model ID)
            temperature: Sampling temperature
            max_tokens: Maximum tokens to generate
            system_prompt: Optional system prompt

        Returns:
            Dict with parsed JSON result or error
        """
        messages = [{"role": "user", "content": prompt}]

        response = self.chat_completion(
            messages=messages,
            model=model,
            temperature=temperature,
            max_tokens=max_tokens,
            system_prompt=system_prompt,
            use_structured_output=True,
            json_schema=json_schema,
        )

        if not response["success"]:
            return response

        try:
            content = response["data"]["choices"][0]["message"]["content"]
            # With structured outputs, content is already valid JSON
            parsed_data = json.loads(content)
            return {
                "success": True,
                "data": parsed_data,
                "model_used": response["model_used"],
            }
        except (KeyError, IndexError) as e:
            return {
                "success": False,
                "error": f"Failed to extract completion: {str(e)}",
            }
        except json.JSONDecodeError as e:
            return {
                "success": False,
                "error": f"Failed to parse JSON response: {str(e)}",
            }

    def list_models(self) -> Dict[str, Any]:
        """
        Get list of available models from OpenRouter API

        Returns:
            Dict with models list or error
        """
        if not self.api_key:
            return {"success": False, "error": "OpenRouter API key not configured"}

        try:
            response = self.network.request(
                {
                    "url": "/v1/models",
                    "method": "GET",
                    "headers": {"Authorization": f"Bearer {self.api_key}"},
                }
            )

            return {"success": True, "models": response["data"]["data"]}

        except NetworkError as e:
            return {"success": False, "error": f"Failed to fetch models: {str(e)}"}

File: tools/communication/openrouter/src/nodejs/openrouter-tool.ts
import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'
import { Network, NetworkError } from '@sdk/network'

// Hardcoded default settings for OpenRouter tool
const OPENROUTER_API_KEY: string | null = null
const OPENROUTER_MODEL = 'google/gemini-3-flash-preview'
const DEFAULT_SETTINGS: Record<string, unknown> = {
  OPENROUTER_API_KEY,
  OPENROUTER_MODEL
}
const REQUIRED_SETTINGS = ['OPENROUTER_API_KEY']

interface ChatMessage {
  role: string
  content: string
}

interface ChatCompletionOptions {
  messages: ChatMessage[]
  model?: string
  temperature?: number
  max_tokens?: number
  system_prompt?: string
  use_structured_output?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json_schema?: Record<string, any>
}

interface CompletionOptions {
  prompt: string
  model?: string
  temperature?: number
  max_tokens?: number
  system_prompt?: string
  use_structured_output?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json_schema?: Record<string, any>
}

interface StructuredCompletionOptions {
  prompt: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json_schema: Record<string, any>
  model?: string
  temperature?: number
  max_tokens?: number
  system_prompt?: string
}

interface ApiResponse {
  success: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  data?: any
  model_used?: string
  error?: string
  status_code?: number
}

export default class OpenRouterTool extends Tool {
  private static readonly TOOLKIT = 'communication'
  private readonly config: ReturnType<typeof ToolkitConfig.load>
  private api_key: string | null
  private model: string
  private readonly network: Network

  constructor(apiKey?: string) {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(OpenRouterTool.TOOLKIT, this.toolName)

    const toolSettings = ToolkitConfig.loadToolSettings(
      OpenRouterTool.TOOLKIT,
      this.toolName,
      DEFAULT_SETTINGS
    )
    this.settings = toolSettings
    this.requiredSettings = REQUIRED_SETTINGS
    this.checkRequiredSettings(this.toolName)

    // Priority: skill-provided apiKey > toolkit settings > hardcoded default
    this.api_key =
      apiKey ||
      (this.settings['OPENROUTER_API_KEY'] as string) ||
      OPENROUTER_API_KEY

    // Load model from toolkit settings or hardcoded default
    this.model =
      (this.settings['OPENROUTER_MODEL'] as string) || OPENROUTER_MODEL

    this.network = new Network({ baseURL: 'https://openrouter.ai/api' })
  }

  get toolName(): string {
    return 'openrouter'
  }

  get toolkit(): string {
    return OpenRouterTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Set the OpenRouter API key
   */
  setApiKey(apiKey: string): void {
    this.api_key = apiKey
  }

  /**
   * Send a chat completion request to OpenRouter
   */
  async chatCompletion(options: ChatCompletionOptions): Promise<ApiResponse> {
    const {
      messages,
      model,
      temperature = 0.7,
      max_tokens,
      system_prompt,
      use_structured_output = false,
      json_schema
    } = options

    if (!this.api_key) {
      return {
        success: false,
        error: 'OpenRouter API key not configured'
      }
    }

    // Use default model if none provided
    const finalModel = model || this.model

    // Prepare messages with system prompt if provided
    const requestMessages = []
    if (system_prompt) {
      requestMessages.push({ role: 'system', content: system_prompt })
    }
    requestMessages.push(...messages)

    // Prepare request payload
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const payload: any = {
      model: finalModel,
      messages: requestMessages,
      temperature
    }

    if (max_tokens) {
      payload.max_tokens = max_tokens
    }

    // Add structured output configuration if requested
    if (use_structured_output && json_schema) {
      payload.response_format = {
        type: 'json_schema',
        json_schema: {
          name: json_schema['name'] || 'response',
          strict: true,
          schema: json_schema['schema']
        }
      }
    }

    try {
      const response = await this.network.request({
        url: '/v1/chat/completions',
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.api_key}`,
          'Content-Type': 'application/json'
        },
        data: payload
      })

      return {
        success: true,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        data: response.data as any,
        model_used: finalModel
      }
    } catch (error: unknown) {
      return {
        success: false,
        error: `OpenRouter API error: ${(error as Error).message}`,
        status_code:
          error instanceof NetworkError ? error.response.statusCode : undefined
      }
    }
  }

  /**
   * General text completion for any use case
   */
  async completion(options: CompletionOptions): Promise<ApiResponse> {
    const {
      prompt,
      model,
      temperature = 0.7,
      max_tokens,
      system_prompt,
      use_structured_output = false,
      json_schema
    } = options

    const messages = [{ role: 'user', content: prompt }]

    const response = await this.chatCompletion({
      messages,
      model: model || this.model,
      temperature,
      max_tokens,
      system_prompt,
      use_structured_output,
      json_schema
    })

    if (!response.success) {
      return response
    }

    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const content = (response.data as any).choices[0].message.content

      return {
        success: true,
        data: { content },
        model_used: response.model_used
      }
    } catch (error: unknown) {
      return {
        success: false,
        error: `Failed to extract completion: ${(error as Error).message}`
      }
    }
  }

  /**
   * Generate structured JSON output using OpenRouter's structured outputs feature
   */
  async structuredCompletion(
    options: StructuredCompletionOptions
  ): Promise<ApiResponse> {
    const {
      prompt,
      json_schema,
      model,
      temperature = 0.7,
      max_tokens,
      system_prompt
    } = options

    const messages = [{ role: 'user', content: prompt }]

    const response = await this.chatCompletion({
      messages,
      model: model || this.model,
      temperature,
      max_tokens,
      system_prompt,
      use_structured_output: true,
      json_schema
    })

    if (!response.success) {
      return response
    }

    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const content = (response.data as any).choices[0].message.content
      const parsedData =
        typeof content === 'string' ? JSON.parse(content) : content

      return {
        success: true,
        data: parsedData,
        model_used: response.model_used
      }
    } catch (error: unknown) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const content = (response.data as any).choices[0]?.message?.content

      if (error instanceof SyntaxError) {
        // Show raw response preview to help debug JSON parsing errors
        const preview =
          typeof content === 'string'
            ? content.substring(0, 500)
            : JSON.stringify(content ?? 'null').substring(0, 500)

        return {
          success: false,
          error: `Failed to parse JSON response: ${error.message}. Response preview: ${preview}`
        }
      } else {
        return {
          success: false,
          error: `Failed to extract completion: ${(error as Error).message}`
        }
      }
    }
  }

  /**
   * Get list of available models from OpenRouter API
   */
  async listModels(): Promise<ApiResponse> {
    if (!this.api_key) {
      return {
        success: false,
        error: 'OpenRouter API key not configured'
      }
    }

    try {
      const response = await this.network.request({
        url: '/v1/models',
        method: 'GET',
        headers: {
          Authorization: `Bearer ${this.api_key}`
        }
      })

      return {
        success: true,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        data: { models: (response.data as any).data }
      }
    } catch (error: unknown) {
      return {
        success: false,
        error: `Failed to fetch models: ${(error as Error).message}`
      }
    }
  }
}
